R/dwdData.R

```diff
@@ -0,0 +1,204 @@
+dwdData <-function(urltble, datasetName, dfile = "all", outdir, sim, module, checkhash = FALSE,
+                    dbHash = "dbHash.sqlite", cascade = FALSE, quick = FALSE){
+
+  # Save environment to recover params in untarDwd and unzipDwd function
+  env = sys.frame(sys.nframe())
+
+  if(missing(urltble)) stop("You must provide a datatable that contain the link
+                            between url, datasetName and password")
+  if(missing(datasetName)) stop("You must provide dataset name to access url of interest")
+
+  setkey(urltble, "dataset")
+  url<- as.character(urltble[.(datasetName)][, 'url', with=FALSE])
+  password<- as.character(urltble[.(datasetName)][, 'password', with=FALSE])
+
+  if(missing(outdir)) {
+    if(!missing(sim)) {
+      if(missing(module)) stop("You must provide a module name, if you provide a simList")
+      outdir <- file.path(modulePath(sim), module, "data")
+    } else {
+      outdir <- file.path(dirname(tempdir()),"data",file_path_sans_ext(basename(url)))
+    }
+  }
+
+  # Create a output dir using datasetName
+  subdir <- file.path(outdir, datasetName, sep="")
+  dir.create(subdir, recursive = TRUE, showWarnings = FALSE)
+
+  # Split url into typeConn(ftp, http) and address.
+  typeConn <- paste(unlist(strsplit(url, "//"))[1],"//", sep ="")
+  address<- unlist(strsplit(url, "//"))[2]
+
+  # Create list of file to download
+  # FTP connection
+  if (typeConn == "ftp://"){
+    if (isTRUE(dfile == "all")){
+      # all files to download. No password for the ftp site
+      if (is.na(password)) { file.list <- getURL(url, ftp.use.epsv = FALSE,
+                                                 dirlistonly = TRUE)
+                            file.list <- paste(url, strsplit(file.list,
+                                                 "\r*\n")[[1]], sep = "")
+      }else{
+        # all files to download. Password needed for the ftp site
+        file.list <- getURL(url, userpwd=password, ftp.use.epsv = FALSE, dirlistonly = TRUE)
+        file.list <- paste(typeConn, password, "@", address,
+                           strsplit(file.list, "\r*\n")[[1]], sep = "")
+      }
+    }else{
+      # Specific files to download on the ftp site no password required
+      if (is.na(password)) {file.list<-paste(url, dfile, sep="")
+      }else{
+        # Specific files to download on the ftp site, password is required
+        file.list<-paste(typeConn, password, "@", address, dfile, sep="")
+      }
+    }
+  }
+  # HTTP connection
+  if (typeConn == "http://"){
+    if (isTRUE(dfile == "all")){
+      file.list <- readHTMLTable(url, skip.rows=1:2)[[1]]$Name
+      file.list <- paste(url, file.list[!is.na(file.list)], sep="")
+    }else{
+      file.list <- paste(url, dfile, sep="")
+    }
+  }
+  # Crosscheck checksum value fromtable where checksum values from previous download are compiled.
```

Comment:
this section feels like a reimplementation of functionality that already exists in *e.g.*, `httr::GET`. See how I'm downloading files already:

- https://github.com/PredictiveEcology/SpaDES/blob/development/R/module-repository.R#L91
- https://github.com/PredictiveEcology/SpaDES/blob/development/R/module-repository.R#L265